### PR TITLE
Some alignment optimizations

### PR DESCRIPTION
**Pull request by benbovy**
- [x] Benchmark added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

May fix some performance regressions, e.g., see https://github.com/pydata/xarray/issues/7376#issuecomment-1352989233.

@ravwojdyla with this PR `ds.assign(foo=~ds["d3"])` in your example should be much faster (on par with version 2022.3.0).

### CODE REVIEW

**Comment by benbovy** [comment]
Quick benchmark taking the example in #7376 (it seems even much faster than in version 2022.3.0!)

```python
# version 2022.3.0
%timeit ds.assign(foo=~ds["d3"])
# 22.5 ms ± 1.96 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

# main branch
%timeit ds.assign(foo=~ds["d3"])
# 193 ms ± 1.35 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# this PR
%timeit ds.assign(foo=~ds["d3"])
# 1.01 ms ± 10.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by Illviljan** [comment]
No obenchmark is catching this? Maybe we can add a small one in https://github.com/pydata/xarray/blob/main/asv_bench/benchmarks/indexing.py ?

File: xarray/core/indexes.py [hunk 1]
```diff
@@ -1419,6 +1419,11 @@ def check_variables():
         )
 
     indexes = [e[0] for e in elements]
+
+    same_objects = all(indexes[0] is other_idx for other_idx in indexes[1:])
```
Comment (Illviljan) [inline xarray/core/indexes.py:1423]:
```suggestion
    indexes_0 = indexes[0]
    same_objects = all(indexes_0 is other_idx for other_idx in indexes[1:])
```
No need to use getitem several times for the same value. Similar thing can be done in other places in the function as well.

Comment (headtr1ck) [reply xarray/core/indexes.py:1423]: List indexing is O(1), so should be only a marginal speedup :P

Comment (benbovy) [reply xarray/core/indexes.py:1423]: Yeah I think the gain in perf would be negligible.

**Comment by benbovy** [comment]
I don't know if the optimizations added here will benefit a large set of use cases (it took 6 months before seeing an issue report), but it is worth for at least a few of them. This is ready I think (added some benchmarks).

**Comment by Illviljan** [comment]
Thanks, @benbovy !

**Pull request merged**